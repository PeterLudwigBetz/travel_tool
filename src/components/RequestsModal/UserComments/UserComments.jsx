
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import ImageLink from '../../image-link/ImageLink';
import ConnectedCommentBox from '../CommentBox/CommentBox';

class UserComments extends Component {
  state = {
    commentToEdit: '',
    activeCommentId: '' 
  };

  editComment = (comment) => {
    const {id}  = comment;
    this.setState ({
      activeCommentId: id,
      commentToEdit: comment.comment,
    });
  }

  resetEditing = () => {
    this.setState({
      activeCommentId: ''
    });
  }

  handleCancelClick = (e) => {
    e.preventDefault();
    this.editComment('');
  }

  formatDate(date) {
    const createdAt = moment(date).format('MM/DD/YYYY @ h:mm a');
    return moment(createdAt).fromNow();
  }
 
  renderDeleteButton () {
    return (
      <span className="modal__dialog">
        <button type="button" className="modal__delete-btn">
            Delete
        </button>
      </span>
    );
    
  }

  renderCancelButton () {
    return (
      <button
        className="editor__post-btn editor__post-btn post-btn-text button-color" id="cancel-button" 
        type="submit" onClick={this.handleCancelClick}>
          Cancel
      </button>
    );

  }

  render() {
    const { commentToEdit, activeCommentId } = this.state;
    const { comments } = this.props;
    return comments && comments.map((comment) => {
      const editedLabel = comment.isEdited  ? '<span>(edited)</span>' : '';
      return (
        <div className="modal__modal1" key={comment.id}>
          <hr />
          <div className="modal__mdl-icons">
            <ImageLink imageSrc={comment.picture} altText="avatar" imageClass="modal__oval-copy" />
            <span className="modal__user-name">
              { comment.userName }
            </span>
            <span className="modal__hours-status">
              {this.formatDate(comment.createdAt)}
            </span>
            <span className="modal__dialog">
              <button 
                type="button" className={`edit-button ${activeCommentId === comment.id ? 'active': ''}`} 
                onClick={() => this.editComment(comment)}>
                Edit
              </button>
            </span>
            {this.renderDeleteButton()}
            {commentToEdit == comment.comment ? (
              <div className="comment-box">
                <ConnectedCommentBox afterSubmit={this.resetEditing} editComment={this.editComment} comment={commentToEdit} requestId={comment.requestId} id={comment.id} />
                {this.renderCancelButton()}
              </div>
            ) :  (
              <div className="modal__modal2">
                  <div className="modal__status-update" dangerouslySetInnerHTML={{ __html: `${comment.comment} ${editedLabel}` }} /> {/*eslint-disable-line*/}
              </div>)} 
          </div>
        </div>
      );
    });
  }
}

UserComments.propTypes = {
  comments: PropTypes.array
};

UserComments.defaultProps = {
  comments: []
};

export default UserComments;
