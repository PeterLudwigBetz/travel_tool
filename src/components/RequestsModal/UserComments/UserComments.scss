.delete-comment-modal {
  width: 243px;
  min-height: 0px;
  height: 120px;
  font-size: 14px;
  border-radius: 2px;
  background-color: #FFFFFF;
  box-shadow: 0 2px 10px 0 rgba(0,0,0,0.15);
  text-align: right;

  .modal-title-bar {
    font-size: 14px;
    font-family: 'DIN Pro';
    height: 30px;
    border-radius: 2px 2px 0 0;
    background-color: #F8F8F8;
    line-height: 18px;
  }

  img {
    height: 10px;
    width: 10px;
  }

  &__text{
    text-align: left;
  }

  &__btn{
    height: 26px;
    width: 82px;
    border-radius: 1px;
    background-color: #FF5359;
    color: #FFFFFF;
  }
}

.delete-modal-overlay {
  position: absolute !important;
  height: 175px;
  width: 335px;
  top: -175px;
  left: -6px;
  outline: none
}

@media screen and (min-width: 420px) and (max-width: 655px) {
  .delete-modal-overlay {
    left: -153px;
  }
}

.blue-text {
  color: #3359db;
  text-decoration: underline;
}

.modal__delete-btn {
  position: relative;
  &:hover{
    color: #3359db;
    text-decoration: underline;
    cursor: pointer;
  }
}