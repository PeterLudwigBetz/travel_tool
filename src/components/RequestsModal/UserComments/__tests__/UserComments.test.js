import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import UserComments from '../UserComments';

const props = {
  comments: [
    {
      id: 1,
      userName: 'Smith Allen',
      picture: '/path/to/image',
      createdAt: '',
      comment: 'Can you clarify why',
      isEdited: true,
      userEmail: 'gdgdgdgdg'
    }
  ],
  email: 'gdgdgdgdg'
};

const formatDateSpy = sinon.spy(UserComments.prototype, 'formatDate');
describe('UserComments component', () => {
  const wrapper = shallow(<UserComments {...props} />);

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call formatDate', () => {
    expect(formatDateSpy.called).toBe(true);
  });

  it('should render the UserComments component as expected', () => {
    expect(wrapper.find('div').length).toBe(4);
    expect(wrapper.find('span').length).toBe(4);
    expect(wrapper.find('ImageLink').length).toBe(1);
    expect(wrapper.find('button').length).toBe(2);
    expect(wrapper.length).toBe(1);
  });

  it('should handle reset editing', () => {
    const wrapper = shallow(<UserComments {...props} />);
    const spy = jest.spyOn(wrapper.instance(), 'resetEditing');
    wrapper.instance().resetEditing();
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should call handleCancel', () => {
    const wrapper = shallow(<UserComments {...props} />);
    const spy = jest.spyOn(wrapper.instance(), 'handleCancelClick');
    const fakeEvent = { preventDefault: () => {} };
    wrapper.setState({
      commentToEdit: 'Can you clarify why'
    });
    const cancelButton = wrapper.find('#cancel-button');
    cancelButton.simulate('click', fakeEvent);
    expect(spy).toHaveBeenCalled();
  });

  it('should submit edited comment', () => {
    const wrapper = shallow(<UserComments {...props} />);
    const spy = jest.spyOn(wrapper.instance(), 'editComment');
    wrapper.state().dataInput = 'comment';
    const editButton = wrapper.find('#edited').at(0);
    editButton.simulate('click');
    expect(spy).toHaveBeenCalled();
  });
});
