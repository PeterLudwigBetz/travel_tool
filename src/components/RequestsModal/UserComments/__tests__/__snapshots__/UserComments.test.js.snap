// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserComments component should match snapshot 1`] = `
<div
  className="modal__modal1"
  key="1"
>
  <hr />
  <div
    className="modal__mdl-icons"
  >
    <ImageLink
      altText="avatar"
      imageClass="modal__oval-copy"
      imageSrc="/path/to/image"
    />
    <span
      className="modal__user-name"
    >
      Smith Allen
    </span>
    <span
      className="modal__hours-status"
    >
      Invalid date
    </span>
    <span
      className="modal__dialog"
    >
      <button
        className="edit-button "
        onClick={[Function]}
        type="button"
      >
        Edit
      </button>
    </span>
    <span
      className="modal__dialog"
    >
      <button
        className="modal__delete-btn"
        type="button"
      >
        Delete
      </button>
    </span>
    <div
      className="modal__modal2"
    >
      <div
        className="modal__status-update"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Can you clarify why ",
          }
        }
      />
       
    </div>
  </div>
</div>
`;
